### PR TITLE
Convert LDAP / Active Directory authentication to use native PHP LDAP…

### PR DESCRIPTION
**Pull request by sunsetsystems**
… instead of adldap2.

I vaguely remember Matrix is using LDAP? If so would be nice to have their input.

This PR comes from my inability to make adldap2 work for a particular client using an OpenLDAP server. Likely I was doing something wrong but the lack of documentation and useful error messages for adldap2 made this solution easier for me.

A benefit here is that you can exclude a list of users from LDAP authentication, making setup and troubleshooting easier.

Also note that you are to put "{login}" into the Distinguished Name field (in Globals) as a placeholder for the OpenEMR username.

For testing you can use the sample LDAP server described here:
  http://www.forumsys.com/tutorials/integration-how-to/ldap/online-ldap-test-server/
Use these values in globals:
  LDAP - Server Name or URI         : ldap.forumsys.com
  LDAP - Distinguished Name of User : uid={login},dc=example,dc=com
  LDAP - Login Exclusions           : administrator (or whatever your admin login is)

Then create a user named "tesla".
When you enable LDAP and log in as that user, the password is "password".

Note the Adldap class can now be removed from the project.

### CODE REVIEW

**Comment by sunsetsystems** [comment]
On the Microsoft side, I don't have an Active Directory server to test with. However this article tells me there's nothing special to be concerned about:

https://www.exchangecore.com/blog/using-ldap-active-directory-authentication-php

So if there are no objections I'll plan to commit this tomorrow.

My remaining question is, what's the "correct" way to remove adldap2? Just "composer remove adldap2/adldap2" and check in the resulting composer.json and composer.lock?

**Comment by matrix-amiel** [comment]
Hi @sunsetsystems 
I've seen your PR just now.
We are happy for new improvements in the AD authentication, but please give us a few days to test it before you merged it into the code, we use AD in our production servers...

**Comment by sunsetsystems** [comment]
Hi @matrix-amiel sounds good, thanks!

**Comment by matrix-amiel** [comment]
Hi @sunsetsystems 
Sorry, but we don't success to connect with your new code.
Anyway, I don't understand why leaving a library with a great interface for old and complex functions?
(we have an idea to add a script that loads all the users from active directory into DB of Openemr and do it with adldap2 is much simpler...).
What are your problems with adldap2? maybe we can help you.

Regarding our test, we try with something like these parameters - 
image
We get an error - "ldap_bind() for user $user failed"
Do you know what is the problem?

File: library/globals.inc.php [hunk 1]
```diff
@@ -1994,32 +1994,29 @@ function gblTimeZones()
...
         ),
-
-        'base_dn' => array(
-            xl('Active Directory - Domains Base'),
+        'gbl_ldap_dn' => array(
+            xl('LDAP - Distinguished Name of User'),
             'text',
             '',
-            xl('Users is the standard windows CN, replace the DC stuff with your domain.')
+            xl('Embed {login} where the OpenEMR login name of the user is to be; for example: uid={login},dc=example,dc=com')
```
Comment (matrix-amiel) [inline library/globals.inc.php]: Why the placeholder of {login} need to be in the input and not in the hard code?

**Comment by sunsetsystems** [comment]
Hi @matrix-amiel 

My reason for this PR, again, is first of all that I could not get the old code to work with the openldap test server ldap.forumsys.com. You might give that a try and see if you can. But also that openldap2 seems overkill when all that is needed is a bind for authentication, especially since it brings in multiple other PHP components for reasons that are not clear.

Regarding AD please read carefully this article that I referenced above:

https://www.exchangecore.com/blog/using-ldap-active-directory-authentication-php

In that example the DN is something like this:

`mydomain\{login}`

... but in any case it's completely dependent on what the DN looks like with your domain controller, and your AD domain is surely not "mydomain". What were your parameters with adldap2?

I don't understand your reference to "{login}" and hard code. In login_operations.php there is this line:

`str_replace('{login}', $user, $GLOBALS['gbl_ldap_dn']),`

where it replaces "{login}" with the OpenEMR username.

Regarding scripting to load users, I'm sure that can be done using ldap_search and related functions without much trouble. See again the article I linked to where they have code that loads in some other parameters.

**Comment by sunsetsystems** [comment]
By the way the error "ldap_bind() for user <user> failed" means that either the DN or the password does not match for the given user. Again I expect it's the DN, it will have to match exactly and that depends on how DNs are set up in your AD domain controller.

File: library/globals.inc.php (hunk 1 above)
Comment (sunsetsystems) [reply library/globals.inc.php]: The placeholder marks where the OpenEMR username will go, to produce the actual user's Distinguished Name for authentication.

Comment (sunsetsystems) [reply library/globals.inc.php]: The placeholder marks where the OpenEMR username will go, to produce the actual user's Distinguished Name for authentication.

**Comment by sunsetsystems** [comment]
Hi @matrix-amiel - Were you able to find out the non-relative Distinguished Name format for your users? I expect "uid=" would not be part of that. If there's some way I can help, let me know.

**Comment by matrix-amiel** [comment]
Hi @sunsetsystems 
Sorry but meanwhile we have errors in the connection to Active Directory in our main environments (UAT and Production), We are testing it a lot of hours but without success.
We still search for a solution with the system department... hopefully we will find but I'm not sure.
So now we can't get your code, I'll update if there is new.

Maybe you prefer to add different global settings for OpenLDAP and Active directory and allow work with both adldap2 and php functions ??

**Comment by matrix-amiel** [comment]
@sunsetsystems 
Good news. All works well.
We need a single minor fix and you can merge the code. In our servers, the user name contains a dot(.) and failed in the Regex in library/authentication/login_operations.php:150, please remove it.

Thank you for your patience...

Amiel

**Comment by sunsetsystems** [comment]
Thank you Amiel! Will do.

**Comment by matrix-amiel** [comment]
Another small question, I didn't understand the preg_match for $pass. Would you be able to explain it?

**Comment by sunsetsystems** [comment]
@matrix-amiel 

It seems that if the password is empty then the LDAP server will attempt an anonymous bind which might succeed, whereas we must make sure the ldap_bind fails if the password is incorrect.

However it seems some LDAP implementations will treat a password starting with a null byte as an empty password, so this code is to detect that situation. I don't recall where I first saw this discussed, but try this article:

https://www.drupal.org/project/ldap/issues/1920094

**Pull request merged**